flake.nix

```diff
@@ -9,24 +9,31 @@
         nixpkgs = self;
       };
 
-      lib = import ./lib;
+      lib = (import ./lib).extend libVersionInfoOverlay;
+
+      libVersionInfoOverlay = finalLib: prevLib: {
+        trivial = prevLib.trivial // {
+          versionSuffix =
+            ".${finalLib.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101")}.${self.shortRev or "dirty"}";
+          version = finalLib.trivial.release + finalLib.trivial.versionSuffix;
+          revisionWithDefault = default: self.rev or default;
+        };
+      };
```

Comment:
This logic should also apply to the `lib/flake.nix` flake, or ideally be moved.

Reply:
> lib/flake.nix flake, or ideally be moved.

TIL about `lib/flake.nix`. Will move the implementation there.

Reply:
Hmm... I'm not sure if we want to have a flake-input for nixpkgs (which would be necessary when moving the code to lib's flake.nix).

Other options I see:
* just duplicate it and add a warning on top that the code needs to be kept in sync with lib's flake
* move the lib overlay somehow into lib itself. Doesn't seem too nice to me: until now flakes are an entry-point only to nixpkgs and lib, with that change we'd assume flakes in lib itself which seems wrong to be since lib should be usable in any Nix evaluation, right? Also, wouldn't that give wrong (or misleading) results when evaluating e.g. `self.shortRev` from `lib/flake.nix`?

Reply:
> I'm not sure if we want to have a flake-input for nixpkgs

I don't see a problem with that; only an inconvenience, which is lock file pollution, which is solvable and should be solved on the Nix side.
- https://github.com/NixOS/nix/issues/7730

Oh and subflakes are still broken. That's an actual problem. Never mind.



> * move the lib overlay somehow into lib itself.

We could have a separate "internal" file in `lib/` that takes care of implementing the version setting, without polluting the `lib` attrset. Both flakes could use that file.



> * Also, wouldn't that give wrong (or misleading) results when evaluating e.g. `self.shortRev` from `lib/flake.nix`?

The whole business about versions is error prone if your expectations of it are too high. Not sure how to manage those expectations.